src/main/java/bio/terra/service/dataset/flight/update/DatasetSchemaUpdateValidateModelStep.java

```diff
@@ -32,22 +36,88 @@ public StepResult doStep(FlightContext context) throws InterruptedException, Ret
+                String.join(", ", uniqueTableNames)));
+      }
+    } else {
+      newTableNames = List.of();
     }
+
+    if (DatasetSchemaUpdateUtils.hasColumnAdditions(updateModel)) {
+      List<DatasetSchemaColumnUpdateModel> addColumns = updateModel.getChanges().getAddColumns();
+      List<String> newAndExistingTableNames = ListUtils.union(existingTableNames, newTableNames);
+
+      List<String> missingTables = new ArrayList<>();
+      List<String> duplicateColumns = new ArrayList<>();
+      for (var columnAddition : addColumns) {
+        String tableName = columnAddition.getTableName();
+        if (!newAndExistingTableNames.contains(tableName)) {
+          missingTables.add(tableName);
+          continue;
+        }
+        if (newTableNames.contains(tableName)) {
+          duplicateColumns.addAll(conflictingNewColumns(tableName, columnAddition.getColumns()));
+        } else {
+          duplicateColumns.addAll(
+              conflictingExistingColumns(tableName, dataset, columnAddition.getColumns()));
+        }
+      }
+      if (!missingTables.isEmpty()) {
+        return failsValidation(
+            "Could not find tables to add columns to",
+            List.of(String.format("Missing tables: %s", String.join(", ", missingTables))));
+      }
+      if (!duplicateColumns.isEmpty()) {
+        return failsValidation(
+            "Cannot overwrite existing or to-be-added columns in tables", duplicateColumns);
+      }
+    }
+
     return StepResult.getStepResultSuccess();
   }
 
+  private List<String> conflictingNewColumns(String tableName, List<ColumnModel> newColumns) {
+    TableModel tableModel =
+        updateModel.getChanges().getAddTables().stream()
+            .filter(tm -> tm.getName().equals(tableName))
+            .findFirst()
+            .orElseThrow();
+    List<String> newTableColumnNames =
+        tableModel.getColumns().stream().map(ColumnModel::getName).collect(Collectors.toList());
+    List<String> newColumnNames = newColumnNames(newColumns);
+    return prependTableName(tableName, ListUtils.intersection(newTableColumnNames, newColumnNames));
+  }
+
+  private List<String> conflictingExistingColumns(
+      String tableName, Dataset dataset, List<ColumnModel> newColumns) {
+    DatasetTable table = dataset.getTableByName(tableName).orElseThrow();
+    List<String> existingColumnNames =
+        table.getColumns().stream().map(Column::getName).collect(Collectors.toList());
+    List<String> newColumnNames = newColumnNames(newColumns);
+    return prependTableName(tableName, ListUtils.intersection(existingColumnNames, newColumnNames));
+  }
+
+  private StepResult failsValidation(String message, List<String> reasons) {
+    return new StepResult(
+        StepStatus.STEP_RESULT_FAILURE_FATAL, new DatasetSchemaUpdateException(message, reasons));
+  }
+
+  private static List<String> newColumnNames(List<ColumnModel> columns) {
```

Comment:
I think this is general enough to move to src/main/java/bio/terra/service/dataset/flight/update/DatasetSchemaUpdateUtils.java, and then we could reuse it in places like https://github.com/DataBiosphere/jade-data-repo/blob/951d02dcdd0b9eec5b3b4ccc2c873fa5f2a3db6a/src/main/java/bio/terra/service/dataset/flight/update/DatasetSchemaUpdateAddColumnsPostgresStep.java#L60